### PR TITLE
Added MH2O chunk

### PR DESCRIPTION
**Pull request by Kaev**

### CODE REVIEW

**Comment by Luzifix** [comment]
Testing with [Kalimdor_25_55.zip](https://github.com/Luzifix/Warcraft.NET/files/3106992/Kalimdor_25_55.zip) failed. The MH2O Chunk look after save different.

**Comment by Kaev** [comment]
The others LVF formats have still to be implemented. Currently LiquidVertexFormat 0 and 2 is supported. LiquidObject support needs DBC reading first.

File: Warcraft.NET/Files/ADT/Chunks/MH2O.cs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+        public const string Signature = "MH2O";
+
+        /// <summary>
+        /// Gets or sets <see cref="MH2OHeader"/>s.
+        /// </summary>
+        public MH2OHeader[] MH2OHeaders { get; set; } = new MH2OHeader[256];
+
+        public MH2O()
+        {
+
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Chunks/MH2O.cs]: Remove whitespace

File: Warcraft.NET/Files/ADT/Chunks/MH2O.cs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    public class MH2O : IIFFChunk, IBinarySerializable
+    {
+        public const string Signature = "MH2O";
+
+        /// <summary>
+        /// Gets or sets <see cref="MH2OHeader"/>s.
+        /// </summary>
+        public MH2OHeader[] MH2OHeaders { get; set; } = new MH2OHeader[256];
+
+        public MH2O()
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Chunks/MH2O.cs]: Missing DocHead

File: Warcraft.NET/Files/ADT/Chunks/MH2O.cs [hunk 3]
```diff
@@ -0,0 +1,180 @@
+﻿using System.IO;
+using Warcraft.NET.Files.Interfaces;
+using Warcraft.NET.Files.ADT.Entrys;
+
+namespace Warcraft.NET.Files.ADT.Chunks
+{
+    public class MH2O : IIFFChunk, IBinarySerializable
+    {
+        public const string Signature = "MH2O";
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Chunks/MH2O.cs]: Missing DocHead

File: Warcraft.NET/Files/ADT/Chunks/MH2O.cs [hunk 4]
```diff
@@ -0,0 +1,180 @@
+﻿using System.IO;
+using Warcraft.NET.Files.Interfaces;
+using Warcraft.NET.Files.ADT.Entrys;
+
+namespace Warcraft.NET.Files.ADT.Chunks
+{
+    public class MH2O : IIFFChunk, IBinarySerializable
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Chunks/MH2O.cs]: Missing DocHead

File: Warcraft.NET/Files/ADT/Chunks/MH2O.cs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        /// <inheritdoc/>
+        public string GetSignature()
+        {
+            return Signature;
+        }
+
+        /// <inheritdoc/>
+        public uint GetSize()
+        {
+            int size = MH2OHeader.GetSize() * 256;
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Chunks/MH2O.cs]:
```suggestion
            return (uint)Serialize().Length;
```

File: Warcraft.NET/Files/ADT/Entrys/MH2OAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace Warcraft.NET.Files.ADT.Entrys
+{
+    public class MH2OAttribute
+    {
+        /// <summary>
+        /// Seems to be useable as visibility information
+        /// </summary>
+        public byte[] Fishable { get; set; } = new byte[8];
+        /// <summary>
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Entrys/MH2OAttribute.cs]: Add missing whitespace

File: Warcraft.NET/Files/ADT/Entrys/MH2OHeader.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace Warcraft.NET.Files.ADT.Entrys
+{
+    public class MH2OHeader
+    {
+        /// <summary>
+        /// Gets or sets the offset of the <see cref="MH2OInstance"/>s.
+        /// </summary>
+        public uint OffsetInstances { get; set; }
+        /// <summary>
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Entrys/MH2OHeader.cs]: Add whitespace

File: Warcraft.NET/Files/ADT/Entrys/MH2OHeader.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        /// Gets or sets the amount of <see cref="MH2OInstance"/>.
+        /// </summary>
+        public uint LayerCount { get; set; }
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Entrys/MH2OHeader.cs]: Add whitespace

File: Warcraft.NET/Files/ADT/Entrys/MH2OHeader.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        /// <summary>
+        /// Gets or sets the offset of the <see cref="MH2OAttribute"/>s.
+        /// </summary>
+        public uint OffsetAttributes { get; set; }
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Entrys/MH2OHeader.cs]: Add whitespace

File: Warcraft.NET/Files/ADT/Entrys/MH2OHeader.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        /// <summary>
+        /// Gets or sets the <see cref="MH2OInstance"/>s of the current <see cref="MH2OHeader"/>.
+        /// </summary>
+        public MH2OInstance[] Instances { get; set; }
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Entrys/MH2OHeader.cs]: Add whitespace

File: Warcraft.NET/Files/ADT/Entrys/MH2OInstance.cs [hunk 11]
```diff
@@ -0,0 +1,109 @@
+﻿using System.IO;
+
+namespace Warcraft.NET.Files.ADT.Entrys
+{
+    public class MH2OInstance
+    {
+        /// <summary>
+        /// Gets or sets the liquid type.
+        /// </summary>
+        public ushort LiquidTypeId { get; set; }
```
Comment (Luzifix) [inline Warcraft.NET/Files/ADT/Entrys/MH2OInstance.cs]: Add whitespace after any attribute

File: Warcraft.NET/Files/Structures/UVMapEntry.cs [hunk 12]
```diff
@@ -0,0 +1,8 @@
+﻿namespace Warcraft.NET.Files.Structures
```
Comment (Luzifix) [inline Warcraft.NET/Files/Structures/UVMapEntry.cs]: Add comments see: https://github.com/Luzifix/Warcraft.NET/blob/master/Warcraft.NET/Files/Structures/Rotator.cs

**Comment by Luzifix** [comment]
[CodeFactor](https://www.codefactor.io/repository/github/luzifix/warcraft.net) found multiple issues last seen at 971e836525e496ac7205772b3ee9a3d47834d242:

#### Insert parentheses within the arithmetic expression to declare the operator precedence.
[Warcraft.NET\Files\ADT\Chunks\MH2O.cs:50
](https://github.com/Luzifix/Warcraft.NET/blob/971e836525e496ac7205772b3ee9a3d47834d242/Warcraft.NET/Files/ADT/Chunks/MH2O.cs#L50)[Warcraft.NET\Files\ADT\Chunks\MH2O.cs:138
](https://github.com/Luzifix/Warcraft.NET/blob/971e836525e496ac7205772b3ee9a3d47834d242/Warcraft.NET/Files/ADT/Chunks/MH2O.cs#L138)[Warcraft.NET\Files\ADT\Chunks\MH2O.cs:91
](https://github.com/Luzifix/Warcraft.NET/blob/971e836525e496ac7205772b3ee9a3d47834d242/Warcraft.NET/Files/ADT/Chunks/MH2O.cs#L91)[Warcraft.NET\Files\ADT\Entrys\MH2OInstanceVertexData.cs:90
](https://github.com/Luzifix/Warcraft.NET/blob/971e836525e496ac7205772b3ee9a3d47834d242/Warcraft.NET/Files/ADT/Entrys/MH2OInstanceVertexData.cs#L90)[Warcraft.NET\Files\ADT\Entrys\MH2OInstance.cs:84
](https://github.com/Luzifix/Warcraft.NET/blob/971e836525e496ac7205772b3ee9a3d47834d242/Warcraft.NET/Files/ADT/Entrys/MH2OInstance.cs#L84)
#### A closing curly bracket must not be preceded by a blank line.
[Warcraft.NET\Files\ADT\Chunks\MH2O.cs:19
](https://github.com/Luzifix/Warcraft.NET/blob/971e836525e496ac7205772b3ee9a3d47834d242/Warcraft.NET/Files/ADT/Chunks/MH2O.cs#L19)
#### An opening curly bracket must not be followed by a blank line.
[Warcraft.NET\Files\ADT\Chunks\MH2O.cs:17
](https://github.com/Luzifix/Warcraft.NET/blob/971e836525e496ac7205772b3ee9a3d47834d242/Warcraft.NET/Files/ADT/Chunks/MH2O.cs#L17)
#### Complex Method
[Warcraft.NET\Files\ADT\Chunks\MH2O.cs:106-179
](https://github.com/Luzifix/Warcraft.NET/blob/971e836525e496ac7205772b3ee9a3d47834d242/Warcraft.NET/Files/ADT/Chunks/MH2O.cs#L106-L179)[Warcraft.NET\Files\ADT\Entrys\MH2OInstanceVertexData.cs:29-87
](https://github.com/Luzifix/Warcraft.NET/blob/971e836525e496ac7205772b3ee9a3d47834d242/Warcraft.NET/Files/ADT/Entrys/MH2OInstanceVertexData.cs#L29-L87)[Warcraft.NET\Files\ADT\Entrys\MH2OInstanceVertexData.cs:94-143
](https://github.com/Luzifix/Warcraft.NET/blob/971e836525e496ac7205772b3ee9a3d47834d242/Warcraft.NET/Files/ADT/Entrys/MH2OInstanceVertexData.cs#L94-L143)

**Pull request merged**